File: balance hidrico regional/TemezRegional.m
function [ETR QC I H V Asup Asub]=TemezRegional(AD,P,ETP) %, P, ETP, QM)
% Modelo INYPSA

Hmax = AD*1.20; 
C = 0.94; 
Imax = 91;
alfa = 0.99 ;

% Modelo pequeñas Presas IMFIA
%{
Hmax = AD*0.92; 
C = 0.30; 
Imax = 386;
alfa = 0.0775;
%}

P=[P(1:12);P];
ETP=[ETP(1:12);ETP];



% condicion inicial
    delta=Hmax-0+ETP(1);
    Po=C*(Hmax-0);

    if P(1)<=Po
        T(1)=0;
    else
        T(1)=((P(1)-Po)^2)/(P(1)+delta-2*Po);
    end

    ETR(1)=min(ETP(1),0+P(1)-T(1));
    H(1)=max(0,0+P(1)-T(1)-ETP(1));


    I(1)=Imax*T(1)/(T(1)+Imax);
    
    Asup(1)=T(1)-I(1);

    Qsub(1)=0*exp(-alfa)+alfa*I(1)*exp(-alfa/2);
    
    V(1)=Qsub(1)/alfa;
    
    Asub(1)=0-V(1)+I(1);

    QC(1)=Asup(1)+Asub(1);

    for t=2:length(P)
        
        
        delta=Hmax-H(t-1)+ETP(t);
        Po=C*(Hmax-H(t-1));
        if P(t)<=Po
            T(t)=0;
        else
            T(t)=((P(t)-Po)^2)/(P(t)+delta-2*Po);
        end

        ETR(t)=min(ETP(t),H(t-1)+P(t)-T(t));
        H(t)=max(0,H(t-1)+P(t)-T(t)-ETP(t));

        I(t)=Imax*T(t)/(T(t)+Imax);

        Asup(t)=T(t)-I(t);
        
        Qsub(t)=Qsub(t-1)*exp(-alfa)+alfa*I(t)*exp(-alfa/2);
        V(t)=Qsub(t)/alfa;
        
        Asub(t)=V(t-1)-V(t)+I(t);

        QC(t)=Asup(t)+Asub(t);
    end





ETR=ETR';
QC=QC';
Asup=Asup';
Asub=Asub';
V=V';
I=I';
H=H';

ETR(1:12)=[];
QC(1:12)=[];
H(1:12)=[];

File: Status_Outlook_Bulletin/notebook/9_Validacion_ESP_jupyter_notebook/TemezRegional.py
import numpy as np

def TemezRegional(AD, P, ETP, H0=0.0, V0=0.0):
    # Modelo INYPSA
    Hmax = AD * 1.20
    C = 0.94
    Imax = 91
    alfa = 0.99

    # Modelo pequeñas Presas IMFIA
    """
    Hmax = AD * 0.92
    C = 0.30
    Imax = 386
    alfa = 0.0775
    """
    T = np.zeros_like(P)
    ETR = np.zeros_like(P)
    H = np.zeros_like(P)
    I = np.zeros_like(P)
    Asup = np.zeros_like(P)
    Qsub = np.zeros_like(P)
    V = np.zeros_like(P)
    Asub = np.zeros_like(P)
    QC = np.zeros_like(P)


    H[0] = H0
    V[0] = V0
    Qsub[0] = alfa * V0

    # condicion inicial
    delta = Hmax - H[0] + ETP[0]
    Po = C * (Hmax - H[0])

    if P[0] <= Po:
        T[0] = 0
    else:
        T[0] = ((P[0] - Po) ** 2) / (P[0] + delta - 2 * Po)

    ETR[0] = min(ETP[0], H[0] + P[0] - T[0])
    H[0] = max(0, H[0] + P[0] - T[0] - ETP[0])

    I[0] = Imax * T[0] / (T[0] + Imax)
    Asup[0] = T[0] - I[0]

    Qsub[0] = Qsub[0] * np.exp(-alfa) + alfa * I[0] * np.exp(-alfa / 2)
    V[0] = Qsub[0] / alfa
    Asub[0] = I[0] - V[0] + V0
    QC[0] = Asup[0] + Asub[0]
    
    for t in range(1, len(P)):
        delta = Hmax - H[t - 1] + ETP[t]
        Po = C * (Hmax - H[t - 1])

        if P[t] <= Po:
            T[t] = 0
        else:
            T[t] = ((P[t] - Po) ** 2) / (P[t] + delta - 2 * Po)

        ETR[t] = min(ETP[t], H[t - 1] + P[t] - T[t])
        H[t] = max(0, H[t - 1] + P[t] - T[t] - ETP[t])
        I[t] = Imax * T[t] / (T[t] + Imax)
        Asup[t] = T[t] - I[t]
        Qsub[t] = Qsub[t - 1] * np.exp(-alfa) + alfa * I[t] * np.exp(-alfa / 2)
        V[t] = Qsub[t] / alfa
        Asub[t] = V[t - 1] - V[t] + I[t]
        QC[t] = Asup[t] + Asub[t]

    # ETR = np.array(ETR)
    # QC = np.array(QC)
    # Asup = np.array(Asup)
    # Asub = np.array(Asub)
    # V = np.array(V)
    # I = np.array(I)
    # H = np.array(H)
    # print(H)
    states = {
        "H": H,
        "V": V,
        "I": I,
        "Asub": Asub
    }

    fluxes = {
        "QC": QC,
        "ETR": ETR,
        "Asup": Asup
    }

    return fluxes, states

